liquid/commitment.go

```diff
@@ -0,0 +1,131 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company
+// SPDX-License-Identifier: Apache-2.0
+
+package liquid
+
+import (
+	"time"
+
+	. "github.com/majewsky/gg/option"
+)
+
+// CommitmentChangeRequest is the request payload format for POST /v1/change-commitments.
+type CommitmentChangeRequest struct {
+	AZ AvailabilityZone `json:"az"`
+
+	// The same version number that was reported in the Version field of a GET /v1/info response.
+	// The liquid shall reject this request if the version here differs from the value in the ServiceInfo currently held by the liquid.
+	// This is used to ensure that Limes does not request commitment changes based on outdated resource metadata.
+	InfoVersion uint64 `json:"infoVersion"`
+
+	// On the first level, the commitment changeset is grouped by project.
+	//
+	// Changesets may span over multiple projects e.g. when moving commitments from one project to another.
+	// In this case, the changeset will show the commitment as being deleted in the source project, and as being created in the target project.
+	ByProject map[ProjectUUID]ProjectCommitmentChangeset `json:"byProject"`
+
+	// Whether Limes allows the liquid to confirm (or deny) this changeset.
+	// If true, Limes will only apply the changeset in its database if the response from the liquid is positive.
+	// If false, Limes has already applied the changeset in its database, and is only notifying the liquid about what happened.
+	//
+	// Examples for ExpectsConfirmation = true include commitments moving into the "confirmed" status, or conversion of commitments between resources.
+	// Examples for ExpectsConfirmation = false include commitments being split or moving into the "expired" status.
+	ExpectsConfirmation bool `json:"expectsConfirmation"`
+}
+
+// CommitmentChangeResponse is the response payload format for POST /v1/change-commitments.
+type CommitmentChangeResponse struct {
+	// If ExpectsConfirmation was true, this field shall be empty if the changeset is confirmed, or contain a human-readable error message if the changeset was rejected.
+	// If ExpectsConfirmation was false, Limes will ignore this field (or, at most, log it silently).
+	RejectionReason string `json:"rejectionReason,omitempty"`
+}
```

Comment:
What about adding more rejection scenarios programmatically, such that it enables Limes to distinguish between transient and permanent errors. Also it allows for more advanced scheduling workflows, especially for future commitments.

Optional: not sure about it, but we could also say which of the commitments could be confirmed and which not. 

```suggestion
type CommitmentChangeResponse struct {
	// If ExpectsConfirmation was true, this field shall be empty if the changeset is confirmed, or contain a human-readable error message if the changeset was rejected.
	// If ExpectsConfirmation was false, Limes will ignore this field (or, at most, log it silently).
	RejectionReason string `json:"rejectionReason,omitempty"`
	// Error code for programmatic handling of rejections.
	// Example values: "InfoVersionMismatch", "FutureCommitmentNotYetSchedulable", "CapacityExceeded"
	ErrorCode string `json:"errorCode,omitempty"`
	// If set, indicates how long Limes should wait before retrying this request.
	// Encoded as a string duration (e.g., "1h", "30m") 
	RetryAfter string `json:"retryAfter,omitempty"`
}
```

Reply:
Re ErrorCode: If this is supposed to change how Limes behaves when receiving this response, I would rather have fields that explicitly say what Limes is supposed to do.

Re RetryAfter: When acting on an API request by a user, this could only be a signal to the user (like a `Retry-After` header) that the user can choose to ignore. Is this fine for you?

The only scenario where a commitment change is not directly in response to an API request by a user is when there is a scheduled state change (e.g. one of `planned -> pending -> confirmed -> superseded/expired`). The only one of those that is rejectable are `pending -> confirmed`. What do you anticipate to be a reasonable RetryAfter value in this situation? Limes will already not retry commitment confirmations faster than once every 15 minutes, and I don't think it's terribly useful to go slower than that.